syntax = "proto3";

package train;

service Trainer {
    rpc RunStep(RunStepReuest) returns (RunStepResponse) {}
    rpc Finish(FinishRequest) returns (FinishResponse) {}
}

message RunStepReuest {
    int32 epoch = 1;
    int32 step = 2;
    repeated Ndarray weights = 3;
}

message RunStepResponse {
    int32 epoch = 1;
    int32 step = 2;
    repeated TFTensor grads = 3;
    bytes loss = 4;
}

enum DataType {
    UNKNOWN = 0;
    FLOAT_16 = 1;
    FLOAT_32 = 2;
    FLOAT_64 = 3;
}

message TFTensor {
    bytes data = 1;
    DataType dtype = 2;
}

message Ndarray {
    bytes data = 1;
    DataType dtype = 2;
    repeated int32 shape = 3;
}

message FinishRequest {}

message FinishResponse {}

// Command to generate pb files.
// python -m grpc_tools.protoc -Iproto --python_out=. --pyi_out=. --grpc_python_out=.  proto/train.proto